uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/PublicHolidayLeaveRequestCreationTest.php

```diff
@@ -531,4 +540,184 @@ public function testCreateLeaveRequestsForAllPublicHolidaysInTheFutureForDefault
+    $publicHoliday2->date = CRM_Utils_Date::processDate('2016-11-06');
+
+    $publicHoliday3 = new PublicHoliday();
+    $publicHoliday3->date = CRM_Utils_Date::processDate('next monday');
+
+    $this->creationLogic->createForAllContacts($publicHoliday);
+    $this->creationLogic->createForAllContacts($publicHoliday2);
+    $this->creationLogic->createForAllContacts($publicHoliday3);
+
+    $expiredBalanceChange1 = $this->getBalanceChangeRecord($balanceChange1->id);
+    $expiredBalanceChange2 = $this->getBalanceChangeRecord($balanceChange2->id);
+
+    //The public holiday date (2016-11-03) is before the date the balance change expired so one will be deducted
+    //from $balanceChange1 and also from $balanceChange2
+    //Public Holiday2 is in past but the date is not before an already expired balance change
+    //Public Holiday3 is in the future, so it does not affect the recalculation
+    $this->assertEquals(-2, $expiredBalanceChange1->amount);
+    $this->assertEquals(-4, $expiredBalanceChange2->amount);
+  }
+
+  public function testExpiredBalanceChangeIsRecalculatedWhenCreatingPublicHolidayWithPastDatesForAContract() {
+    $absencePeriod = AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-06-01'),
+      'end_date' => CRM_Utils_Date::processDate('+10 days')
+    ]);
+
+    $contact1 = ContactFabricator::fabricate();
+
+    $periodEntitlement1 = LeavePeriodEntitlementFabricator::fabricate([
+      'contact_id' => $contact1['id'],
+      'period_id' => $absencePeriod->id,
+      'type_id' => $this->absenceType->id,
+    ]);
+
+    $contract1 = HRJobContractFabricator::fabricate([
+      'contact_id' => $contact1['id']
+    ],
+    [
+      'period_start_date' => CRM_Utils_Date::processDate('2016-06-01'),
+    ]);
+
+    $expiryDate = new DateTime('2016-11-04');
+    $balanceChange1 = $this->createExpiredBroughtForwardBalanceChange(
+      $periodEntitlement1->id,
+      3,
+      3,
+      $expiryDate
+    );
+
+    PublicHolidayFabricator::fabricateWithoutValidation([
+      'date' =>  CRM_Utils_Date::processDate('2016-11-03')
+    ]);
+
+    PublicHolidayFabricator::fabricateWithoutValidation([
+      'date' =>  CRM_Utils_Date::processDate('2016-11-06')
+    ]);
+
+    PublicHolidayFabricator::fabricateWithoutValidation([
+      'date' =>  CRM_Utils_Date::processDate('next monday')
+    ]);
+
+    $this->creationLogic->createAllForContract($contract1['id']);
+
+    $expiredBalanceChange1 = $this->getBalanceChangeRecord($balanceChange1->id);
+    //The public holiday date (2016-11-03) is before the date the balance change expired so one will be deducted
+    //from $balanceChange1
+    //Public Holiday2 is in past but the date is not before an already expired balance change
+    //Public Holiday3 is in the future, so it does not affect the recalculation
+    $this->assertEquals(-2, $expiredBalanceChange1->amount);
+  }
+
+  private function getBalanceChangeRecord($balanceChangeID) {
```

Comment:
well... my idea was to actually replace `getBalanceChangeRecord()` with `findById()`. Since now one method is just delegating the job to the other, there's really no reason for `getBalanceChangeRecord()` to exist